lib/Db/Row2Mapper.php

```diff
@@ -0,0 +1,645 @@
+		if(!$columns) {
+			throw new InternalError(get_class($this) . ' - ' . __FUNCTION__ . ': columns are missing');
+		}
+		$this->setColumns($columns);
+
+		// create a new row sleeve to get a new rowId
+		$rowSleeve = $this->createNewRowSleeve($row->getTableId());
+		$row->setId($rowSleeve->getId());
+
+		// write all cells to its db-table
+		foreach ($row->getData() as $cell) {
+			$this->insertCell($rowSleeve->getId(), $cell['columnId'], $cell['value']);
+		}
+
+		return $row;
+	}
+
+	/**
+	 * @throws InternalError
+	 */
+	public function update(Row2 $row, array $columns): Row2 {
+		if(!$columns) {
+			throw new InternalError(get_class($this) . ' - ' . __FUNCTION__ . ': columns are missing');
+		}
+		$this->setColumns($columns);
+
+		// if nothing has changed
+		if (count($row->getChangedCells()) === 0) {
+			return $row;
+		}
+
+		// update meta data for sleeve
+		try {
+			$sleeve = $this->rowSleeveMapper->find($row->getId());
+			$this->updateMetaData($sleeve);
+			$this->rowSleeveMapper->update($sleeve);
+		} catch (DoesNotExistException|MultipleObjectsReturnedException|Exception $e) {
+			$this->logger->error($e->getMessage(), ['exception' => $e]);
+			throw new InternalError(get_class($this) . ' - ' . __FUNCTION__ . ': '.$e->getMessage());
+		}
+
+		// write all changed cells to its db-table
+		foreach ($row->getChangedCells() as $cell) {
+			$this->insertOrUpdateCell($sleeve->getId(), $cell['columnId'], $cell['value']);
+		}
+
+		return $row;
+	}
+
+	/**
+	 * @throws Exception
+	 */
+	private function createNewRowSleeve(int $tableId): RowSleeve {
+		$rowSleeve = new RowSleeve();
+		$rowSleeve->setTableId($tableId);
+		$this->updateMetaData($rowSleeve, true);
+		return $this->rowSleeveMapper->insert($rowSleeve);
+	}
+
+	/**
+	 * Updates the last_edit_by and last_edit_at data
+	 * optional adds the created_by and created_at data
+	 *
+	 * @param RowSleeve|RowCellSuper $entity
+	 * @param bool $setCreate
+	 * @return void
+	 */
+	private function updateMetaData($entity, bool $setCreate = false): void {
+		$time = new DateTime();
+		if ($setCreate) {
+			$entity->setCreatedBy($this->userId);
+			$entity->setCreatedAt($time->format('Y-m-d H:i:s'));
+		}
+		$entity->setLastEditBy($this->userId);
+		$entity->setLastEditAt($time->format('Y-m-d H:i:s'));
+	}
+
+	/**
+	 * Insert a cell to its specific db-table
+	 *
+	 * @throws InternalError
+	 */
+	private function insertCell(int $rowId, int $columnId, $value): void {
+		$cellClassName = 'OCA\Tables\Db\RowCell'.ucfirst($this->columns[$columnId]->getType());
+		/** @var RowCellSuper $cell */
+		$cell = new $cellClassName();
+
+		$cell->setRowIdWrapper($rowId);
+		$cell->setColumnIdWrapper($columnId);
+		$this->updateMetaData($cell);
+
+		// insert new cell
+		$cellMapperClassName = 'OCA\Tables\Db\RowCell'.ucfirst($this->columns[$columnId]->getType()).'Mapper';
+		/** @var QBMapper $cellMapper */
+		try {
+			$cellMapper = Server::get($cellMapperClassName);
```

Comment:
I feel the logic to get the specific class could be reusable in a separate method